dev/com.ibm.ws.jmx/src/com/ibm/ws/jmx/internal/DelayedMBeanActivator.java

```diff
@@ -52,14 +56,19 @@
 final class DelayedMBeanActivator extends MBeanServerForwarderDelegate implements DelayedMBeanHelper {
 
     public static final String MBEAN_CLASSES = "com.ibm.ws.jmx.delayed.MBeanClasses";
+    //DelayedMBeanActivatorHelper helper = new DelayedMBeanActivatorHelper();
+    //private EventAdmin eventAdmin = helper.getEventAdmin();
 
     private volatile MBeanServerNotificationSupport notificationSupport;
     private final ConcurrentHashMap<ObjectName, DelayedMBeanHolder> delayedMBeanMap;
     private final BundleContext bundleContext;
+    private final EventAdmin eventAdmin;
 
-    DelayedMBeanActivator(BundleContext ctx) {
+    DelayedMBeanActivator(BundleContext ctx, EventAdmin eventAdmin) {
```

Comment:
Is auditing configurable by the user? Is this feature always on? Seems like it's unconditionally adding more traffic / object creation in the MBeanServer. Is there always a consumer for the events it's creating?

com.ibm.ws.kernel.boot.jmx.service.MBeanServerPipeline was designed with future logging/auditing/security in mind. It would probably be cleaner to put the auditing logic in its own MBeanServerForwarderDelegate and inject it at the end of the MBeanServerPipeline when your auditing feature is enabled.

Reply:
Auditing is enabled only when the audit-1.0 feature is added, so yes, it is configurable.  Additionally, unless the user specifies that they want the JMX_MBEAN event, it will not get generated, so that too is configurable.    There is a consumer for the event, AuditServiceImpl which implements the EventHandler amongst other things, that listens on kernel-level events.

I will leave the logic in the current class until which time the JMX team has the time to, and wants to, refactor that code.   As a note, the auditing feature is not part of the kernel so is not available for writing audit records other than thru events so as not to create a dependency on audit in the kernel (no one wants that).

Reply:
I guess my fundamental question is why is this is not isolated to the audit-1.0 feature? It's not evident that this code needed to be here.